sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
-            // from refreshing any caches.
-            forceRefreshAddressCache = false;
+            logger.info("Received invalid collection exception, will retry, {}", exception.toString());
+            this.request.forceNameCacheRefresh = true;
+
+            return Pair.of(null, false);
         }
-        return Mono.just(ShouldRetryResult.retryAfter(backoffTime,
-                Quadruple.with(forceRefreshAddressCache, true, timeout, currentRetryAttemptCount)));
     }
 
-    private void stopStopWatch(StopWatch stopwatch) {
-        synchronized (stopwatch) {
-            stopwatch.stop();
+    static class RetryWithRetryPolicy extends RetryPolicyWithDiagnostics {
+        private final static int DEFAULT_WAIT_TIME_IN_SECONDS = 30;
+        private final static int MAXIMUM_BACKOFF_TIME_IN_MS = 15000;
+        private final static int INITIAL_BACKOFF_TIME_MS = 10;
+        private final static int BACK_OFF_MULTIPLIER = 2;
+
+        private final RxDocumentServiceRequest request;
+        private volatile int attemptCount = 1;
+        private volatile int currentBackoffMilliseconds = RetryWithRetryPolicy.INITIAL_BACKOFF_TIME_MS;
+
+        private final int waitTimeInSeconds;
+        private final StopWatch durationTimer;
+        //TODO once this is moved to IRetryPolicy, remove from here.
+        public final static Quadruple<Boolean, Boolean, Duration, Integer> INITIAL_ARGUMENT_VALUE_POLICY_ARG = Quadruple.with(false, false,
+            Duration.ofSeconds(60), 0);
+
+        public RetryWithRetryPolicy(RxDocumentServiceRequest request,
+                                    Integer waitTimeInSeconds,
+                                    StopWatch durationTimer) {
+            this.request = request;
+            this.waitTimeInSeconds = waitTimeInSeconds != null ? waitTimeInSeconds : DEFAULT_WAIT_TIME_IN_SECONDS;
+            this.durationTimer = durationTimer;
         }
-    }
 
-    private void startStopWatch(StopWatch stopwatch) {
-        synchronized (stopwatch) {
-            stopwatch.start();
+        @Override
+        public Mono<ShouldRetryResult> shouldRetry(Exception exception) {
+            Duration backoffTime;
+            Duration timeout;
+
+            if (!(exception instanceof RetryWithException)) {
+                logger.debug("Operation will NOT be retried. Current attempt {}, Exception: ", this.attemptCount,
+                    exception);
+                return Mono.just(ShouldRetryResult.noRetry());
+            }
+
+            RetryWithException lastRetryWithException = (RetryWithException)exception;
+            this.request.setLastRetryWithException(lastRetryWithException);
+
+            long remainingMilliseconds = (this.waitTimeInSeconds * 1000) - this.durationTimer.getTime();
+            int currentRetryAttemptCount = this.attemptCount++;
+
+            if (remainingMilliseconds <= 0) {
+                logger.debug("Received retrywith exception after backoff/retry. Will fail the request.",
```

Comment:
Do the level needs to be higher like at-least WARN?

Reply:
Agreed - Fixed in next iteration.